htlcswitch/link.go

```diff
@@ -538,15 +627,186 @@ func (l *channelLink) WaitForShutdown() {
+		}
+	}
+
+	return nil
+}
+
+// DiableAdds sets the ChannelUpdateHandler state to allow UpdateAddHtlc's in
+// the specified direction. It returns an error if the state already disallowed
+// those adds.
+func (l *channelLink) DisableAdds(linkDirection LinkDirection) error {
+	if linkDirection == Outgoing {
+		if l.isOutgoingAddBlocked.Swap(true) {
+			return errors.New("outgoing adds already disabled")
+		}
+	}
+
+	if linkDirection == Incoming {
+		if l.isIncomingAddBlocked.Swap(true) {
+			return errors.New("incoming adds already disabled")
+		}
+	}
+
+	return nil
+}
+
+// IsFlushing returns true when UpdateAddHtlc's are disabled in the direction of
+// the argument.
+func (l *channelLink) IsFlushing(linkDirection LinkDirection) bool {
+	if linkDirection == Outgoing {
+		return l.isOutgoingAddBlocked.Load()
+	}
+
+	return l.isIncomingAddBlocked.Load()
+}
+
+// OnFlushedOnce is a method that adds a hook to be called the next time the
+// channel state reaches zero htlcs. This hook will only ever be called once.
+// This function returns a FlushHookID that can be used to preemptively remove
+// this hook. If the channel state is already flushed -- no HTLCs and no
+// commitments owed by either party -- then we will call this hook synchronously
+// and will not return a FlushHookID.
+func (l *channelLink) OnFlushedOnce(hook func()) fn.Option[FlushHookID] {
+	// If the channel state is clean, meaning there are no HTLCs, and both
+	// peers have irrevocably committed that state, then we can call this
+	// hook synchronously and not add it to the map at all.
+	if l.channel.IsChannelClean() {
+		hook()
+		return fn.None[FlushHookID]()
+	}
+
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	return fn.Some(FlushHookID(l.flushHooks.alloc(hook, false)))
+}
+
+// OnFlushedMany is a method that adds a hook to be called every time the
+// channel state reaches zero htlcs. This function returns a FlushHookID that
+// can be used to preemptively remove this hook.
+func (l *channelLink) OnFlushedMany(hook func()) FlushHookID {
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	return FlushHookID(l.flushHooks.alloc(hook, true))
+}
+
+// RemoveFlushHook will remove a hook previously added by OnFlushedOnce or
+// OnFlushedMany. It returns an error if the supplied FlushHookID was not found.
+func (l *channelLink) RemoveFlushHook(hid FlushHookID) error {
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	if l.flushHooks.delete(uint64(hid)) {
+		return nil
+	}
+
+	return errors.New("could not remove flush hook: not found")
+}
+
+// OnCommitOnce adds a hook that will be called the next time a CommitSig
+// message is sent in the argument's LinkDirection. This hook will only ever be
+// called once. This function returns a CommitHookID that can be used to
+// preemptively remove this hook. If no CommitSig is owed in the argument's
+// LinkDirection, then we will call this hook synchronously and will not return
+// a CommitHookID.
+func (l *channelLink) OnCommitOnce(
```

Comment:
Similar to `OnCommitOnce`, the `hook` function may never get called:

- goroutine 1: `OnCommitOnce` called, `OweCommitment` is true so `runSync` is false
- goroutine 2: we send `commit_sig` covering their updates on their commitment tx and come to a clean state, but there are no `outgoingCommitHooks` so the `invoke` call is a no-op
- goroutine 1: `outgoingCommitHooks.alloc` is called
- goroutine 1: `OnFlushedOnce` called and since the channel is clean, `hook` is run

In this case, `closing_signed` is sent without us sending a `shutdown`